### PR TITLE
Triplet Loss Function for Siamese Network (issue #1699)

### PR DESCRIPTION
**Pull request by Hemal-Mamtora**
@zoq @shardulparab97 Please have a look at the implementation code of the Triplet Loss Function, Please revert if modification is required.
Thank you

### CODE REVIEW

**Comment by zoq** [comment]
Thanks for the contribution, before I'll take a closer look, can you write a test for the loss function, see https://github.com/mlpack/mlpack/blob/master/src/mlpack/tests/loss_functions_test.cpp for some examples.

**Comment by zoq** [comment]
@mlpack-jenkins test this please

**Comment by Hemal-Mamtora** [comment]
@zoq I had some compile-time errors when I wrote the tests. I will start working again on them from the next week.

**Comment by zoq** [comment]
Sounds good, let me know if you need any help.

File: src/mlpack/methods/ann/loss_functions/triplet_loss_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    const InputType&& negative)
+{
+  arma::mat positive_distance = 
+      metric::SquaredEuclideanDistance::Evaluate(anchor, positive);
+  
+  arma::mat negative_distance = 
+       metric::SquaredEuclideanDistance::Evaluate(anchor, negative);
+
+  double triplet_loss = 
+      std::max(positive_distance - negative_distance + margin, 0);
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/triplet_loss_impl.hpp]: `std::max` doesn't work on `area::mat` (`positive_distance, negative_distance`)`, not sure if that is right, but you could take sum (`area::accu`) over `positive_distance - negative_distance + margin` to solve the issue.

**Comment by johnsoncarl** [comment]
Hey @zoq @hemam-mamtora , I found this too, while I was trying to write loss function test for this. 

I've reported this issue in the IRC channel yesterday, in the main channel and to you in DM.

Though I found it was not a compile time error but was encountered while running an example. I tried using colvec instead of mat, that thing didn't work too. 

I didn't give it much time as the build process eats so much time, despite turning of few not required builts. 

Would you like me to have a look at it too. Or I shall leave it to @hemam-mamtora ?

File: src/mlpack/methods/ann/loss_functions/triplet_loss_impl.hpp (hunk 1 above)
Comment (johnsoncarl) [reply src/mlpack/methods/ann/loss_functions/triplet_loss_impl.hpp]: Yeah, I realised this too, that accumulation function would help here. 

I will be doing this today later.

**Comment by Hemal-Mamtora** [comment]
There was an issue on my local machine, so i pulled from mlpack/mlpack and then merged my branch with it. Apparently this has caused all other pull request to be included in my pull request, so I have closed this pull request, I will open a new one, with the required changes.
